server/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -593,6 +603,46 @@ public void prepareForClose() {
         } catch (Exception e) {
             logger.warn("unexpected exception while waiting for http server to close", e);
         }
+
+        // Wait for async search tasks to finish. Since the HTTP server is closed, none of these tasks should be for synchronous searches
+        // even though the action name is the same. This is a stopgap solution (read: dirty hack), we should figure out a way to
+        // manage Tasks at shutdown time more generally.
+        var taskManager = injector.getInstance(TransportService.class).getTaskManager();
+        TimeValue asyncSearchTimeout = SHUTDOWN_SEARCH_TIMEOUT_SETTING.get(this.settings());
+        FutureTask<Void> searchAwaiter = new FutureTask<>(() -> {
+            boolean searchTasksRemaining = true;
+            while (searchTasksRemaining) {
+                searchTasksRemaining = taskManager.getTasks()
+                    .values()
+                    .stream()
+                    .anyMatch(task -> SearchAction.INSTANCE.name().equals(task.getAction())) == false;
+                if (searchTasksRemaining) {
```

Comment:
Instead of using anyMatch, perhaps use count so we can log (at debug level) the number of tasks remaining